### DISCUSSION TITLE
Additional UnitCodes based on rec20-Rev15-2020

### DISCUSSION THREAD

**Discussion by hh-it-co** [stephanstapel/ZUGFeRD-csharp#365] [Ideas]
Since i need the UnitCode MWH, which is not included in the recommdations, but also listet here https://unece.org/sites/default/files/2023-10/rec20_Rev15e-2020.xls and here
https://docs.peppol.eu/poacc/billing/3.0/codelist/UNECERec20/
IMHO it would make sense to implement the complete list of UnitCodes to be more flexible for different industries (in this case wholesome energy trading). I may create a pull request for this.

**Comment by stephanstapel** [comment]
I once had the complete code lists in the library but removed them because this made the library almost unusable because you couldn't find anything. The lists are simply too long. If you need a certain value, feel free to send a PR for this specific value

**Comment by hh-it-co** [comment]
Ah ok, i also thought, that might ge a reason, why the list is so short.
You can use this code to create a "complete" QuantityCodes.cs (with each usable UnitCode):

`using System.Text;
using NPOI.HSSF.UserModel; // NPOI for reading .xls files
using NPOI.SS.UserModel;

class Program
{
    static void Main()
    {
        // Path to the Excel file
        string filePath = "rec20_Rev15e-2020.xls";

        // Create a FileStream to read the Excel file
        using (FileStream file = new FileStream(filePath, FileMode.Open, FileAccess.Read))
        {
            // Initialize a workbook object to read the .xls file
            HSSFWorkbook workbook = new HSSFWorkbook(file);

            // Look for the sheet named "Annex II & Annex III"
            ISheet sheet = workbook.GetSheet("Annex II & Annex III");

            // Check if the specified sheet was found
            if (sheet == null)
            {
                Console.WriteLine("The sheet 'Annex II & Annex III' was not found.");
                return; // Exit if the sheet does not exist
            }

            // Initialize a StringBuilder to construct the enum in C#
            StringBuilder enumBuilder = new StringBuilder();
            enumBuilder.AppendLine("public enum QuantityCodes");
            enumBuilder.AppendLine("{");

            // Log the total number of rows in the sheet
            Console.WriteLine($"Total rows: {sheet.LastRowNum}");

            // Start reading from the row where the data for codes begins
            int startRow = 10; // Make sure this is the correct starting row
            for (int row = startRow; row <= sheet.LastRowNum; row++)
            {
                // Get the current row from the sheet
                IRow currentRow = sheet.GetRow(row);

                // Check if the current row is not null
                if (currentRow != null)
                {
                    // Get the cell that indicates status (first cell)
                    ICell statusCell = currentRow.GetCell(0);

                    // Only process if the status cell does not contain "X"
                    if (statusCell.ToString() != "X")
                    {
                        // Read the cell for the quantity code (Column A)
                        ICell codeCell = currentRow.GetCell(1); // Column A: Quantity Code
                        string? code = codeCell?.ToString();

                        // Check if the first character of the code is a digit
                        if (!int.TryParse(code?.First().ToString(), out var result))
                        {
                            // Read the cell for the description (Column B)
                            ICell descriptionCell = currentRow.GetCell(2); // Column B: Description
                            string? description = descriptionCell?.ToString();

                            // Log the current row's code and description
                            Console.WriteLine($"Row {row}: Code = {code}, Description = {description}");

                            // Check if both the code and description are valid
                            if (!string.IsNullOrWhiteSpace(code) && !string.IsNullOrWhiteSpace(description))
                            {
                                // Clean the code and description
                                code = code.Trim();
                                description = description.Trim().Replace("\"", ""); // Remove quotes

                                // Create a valid enum name from the code
                                string enumName = code.Replace(" ", "_").Replace("-", "_").Replace("/", "_");

                                // Add the enum element with its description
                                enumBuilder.AppendLine($"\t/// <summary>");
                                enumBuilder.AppendLine($"\t/// {description}");
                                enumBuilder.AppendLine($"\t/// </summary>");
                                enumBuilder.AppendLine($"\t{enumName},");
                            }
                        }
                    }
                }
            }

            enumBuilder.AppendLine("}"); // Close the enum definition

            // Save the generated enum to a .cs file
            string outputPath = "QuantityCodes.cs";
            File.WriteAllText(outputPath, enumBuilder.ToString());

            // Confirm that the enum file has been successfully generated
            Console.WriteLine("Enum file has been successfully generated: " + outputPath);
        }
    }
}
`

There might be a solution, to use such huge enums in a handier way. Maybe the codes can be grouped and devided into smaller enums or so?
Another solution could be, to enable a free-text-unit/quantity-code.
Another solution could be, to use classes instead of an enum. With a base class or an interface, that requires a code and a description, you may enable users to add their own quantity-codes by inheritating the base-classs/interface.

For me, it would be enough to add MWH.

**Comment by hh-it-co** [comment]
I created the pull request https://github.com/stephanstapel/ZUGFeRD-csharp/pull/366
